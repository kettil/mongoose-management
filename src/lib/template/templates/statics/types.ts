/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/naming-convention */
/*
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

import { IndexOptions } from 'mongodb';
import { Document, DocumentQuery, Aggregate, Model } from 'mongoose';

// ##### Virtuals #####

/**
 * Helpers
 */
type Keys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type Diff<T, U> = T extends U ? never : T;

/**
 * Extracts the type from the function
 */
type VirtualFunctionTypeGet<T extends { [k: string]: any }> = T['get'] extends () => void
  ? ReturnType<T['get']>
  : never;
type VirtualFunctionTypeSet<T extends { [k: string]: any }> = T['set'] extends (value: infer A) => void ? A : never;

/**
 * List of all types (with the type "never")
 */
type VirtualFunctionTypeGetList<T> = { [K in keyof T]: VirtualFunctionTypeGet<T[K]> };
type VirtualFunctionTypeSetList<T> = { [K in keyof T]: VirtualFunctionTypeSet<T[K]> };

/**
 * List of all types (without the type "never")
 */
type VirtualTypeGetList<T> = Pick<VirtualFunctionTypeGetList<T>, Keys<VirtualFunctionTypeGetList<T>>>;
type VirtualTypeSetList<T> = Pick<VirtualFunctionTypeSetList<T>, Keys<VirtualFunctionTypeSetList<T>>>;

/**
 * List of Keys where the object have only "get"
 */
type VirtualKeysGetOnly<T> = Diff<keyof VirtualTypeGetList<T>, keyof VirtualTypeSetList<T>>;

/**
 * List of Keys where without "get only" objects
 */
type VirtualKeysBoth<T> = Diff<keyof T, VirtualKeysGetOnly<T>>;

type VirtualSet<T> = { [K in keyof VirtualTypeSetList<T>]: VirtualTypeSetList<T>[K] };
type VirtualGet<T> = { [K in keyof VirtualTypeGetList<T> & VirtualKeysBoth<T>]: VirtualTypeGetList<T>[K] };
type VirtualOnlyGet<T> = {
  readonly [K in keyof VirtualTypeGetList<T> & VirtualKeysGetOnly<T>]: VirtualTypeGetList<T>[K];
};

export type VirtualType<T> = VirtualOnlyGet<T> & VirtualGet<T> & VirtualSet<T>;

// ##### Hooks #####

type HooksMethodDocument = 'validate' | 'save' | 'remove';
type HooksMethodQueryWithout = 'update' | 'updateOne' | 'updateMany' | 'deleteOne' | 'deleteMany';
type HooksMethodQueryCount = 'countDocuments';
type HooksMethodQuerySingle = 'findOne' | 'findOneAndDelete' | 'findOneAndRemove' | 'findOneAndUpdate';
type HooksMethodQueryMultiple = 'find';
type HooksMethodAggreagate = 'aggregate';
type HooksMethodInsert = 'insertMany';

type HooksReturn = Promise<void> | void;

type HooksPre<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  A extends Aggregate<D>,
  M extends Model<D>,
  E extends Record<string, unknown>
> = {
  (method: 'init', cb: (this: D) => void): void;
  (method: HooksMethodDocument, cb: (this: D) => HooksReturn): void;
  (method: HooksMethodQueryCount, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodQueryMultiple, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodQuerySingle, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodQueryWithout, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodAggreagate, cb: (this: A) => HooksReturn): void;
  (method: HooksMethodInsert, cb: (this: M) => HooksReturn): void;
};

type HooksPost<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  A extends Aggregate<D>,
  M extends Model<D>,
  E extends Record<string, unknown>
> = {
  (method: 'init', cb: (this: D, doc: D) => void): void;
  (method: HooksMethodDocument, cb: (this: D, doc: D) => HooksReturn): void;
  (method: HooksMethodQueryCount, cb: (this: Q & E, count: number) => HooksReturn): void;
  (method: HooksMethodQueryMultiple, cb: (this: Q & E, docs: D[]) => HooksReturn): void;
  (method: HooksMethodQuerySingle, cb: (this: Q & E, doc: D) => HooksReturn): void;
  (method: HooksMethodQueryWithout, cb: (this: Q & E) => HooksReturn): void;
  (method: HooksMethodAggreagate, cb: (this: A, docs: D[]) => HooksReturn): void;
  (method: HooksMethodInsert, cb: (this: M, docs: D[]) => HooksReturn): void;
};

type ClearExtendQuery<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  E extends Record<string, unknown>
> = Partial<Omit<E, keyof Q>>;

export type HooksHandler<
  D extends Document,
  Q extends DocumentQuery<D | D[], D>,
  A extends Aggregate<D>,
  M extends Model<D>,
  ExtendQuery extends Record<string, unknown>
> = (
  pre: HooksPre<D, Q, A, M, ClearExtendQuery<D, Q, ExtendQuery>>,
  post: HooksPost<D, Q, A, M, ClearExtendQuery<D, Q, ExtendQuery>>,
) => void;

// ##### Others #####

export type ExtendTimestampType = {
  readonly createdAt: Date;
  readonly updatedAt: Date;
};

export type ExtendIdType = {
  id: any;
};

export type ExportType<T extends Record<string, unknown>> = T &
  ExtendTimestampType &
  ExtendIdType & {
    _id: unknown;
    __v: unknown;
  };

type IndexOptionsExclude = 'dropDups' | 'min' | 'max' | 'v' | 'session' | 'w' | 'j' | 'wtimeout';

export type IndexType = {
  fields: { [k: string]: 1 | -1 | 'text' | '2dsphere' | 'hashed' };
  options: Pick<IndexOptions, Exclude<keyof IndexOptions, IndexOptionsExclude>>;
};
