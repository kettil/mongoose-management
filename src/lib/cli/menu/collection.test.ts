jest.mock('../../prompts');

import Prompts from '../../prompts';
import CollectionDataset from '../dataset/collection';
import ColumnMenu from './column';

import CollectionMenu from './collection';

import { choicesType } from '../../types';

describe('Check the CollectionMenu class', () => {
  let collection: CollectionDataset;
  let prompts: any;
  let menu: any;

  beforeEach(() => {
    collection = new CollectionDataset(
      {
        name: 'c1',
        columns: [{ name: 'column1', type: 'string' }],
        indexes: [{ name: 'index1', columns: { column1: 1 }, properties: {} }],
        idType: 'uuidv4',
      },
      jest.fn() as any,
    );
    collection.setReference();

    prompts = new Prompts();

    menu = new CollectionMenu(prompts);
  });

  test('initialize the class', () => {
    expect(menu).toBeInstanceOf(CollectionMenu);

    expect(menu.prompts).toBe(prompts);
    expect(menu.columnMenu).toBeInstanceOf(ColumnMenu);
  });

  test('it should be return menu selection when exec() is called', async () => {
    ((prompts.menu as any) as jest.Mock).mockImplementation((_: string, choices: choicesType<any>[]) => {
      expect(choices).toMatchSnapshot();

      return { action: 'edit', data: undefined };
    });

    const result = await menu.exec(collection);

    expect(result).toEqual({ action: 'edit', data: undefined });
    expect(prompts.menu).toHaveBeenCalledTimes(1);
    expect(prompts.menu).toHaveBeenCalledWith('Choose a column/index or a command:', expect.any(Array));
  });

  test('it should be return menu selection when getChoiceIndexList() is called with indexes', () => {
    collection.getColumn('column1')!.setIndex(1, 'index');

    const result = menu.getChoiceIndexList(collection.getIndexes());

    expect(result).toMatchSnapshot();
  });

  test('it should be return menu selection when getChoiceIndexList() is called without indexes', () => {
    collection.getIndexes().forEach((i) => i.remove());

    const result = menu.getChoiceIndexList(collection.getIndexes());

    expect(result).toMatchSnapshot();
  });

  test.each<[any]>([['index'], ['unique'], ['sparse']])(
    'it should be return indexes table when createIndexTable() is called with type "%s"',
    (type) => {
      collection.getColumn('column1')!.setIndex(1, type);

      const result = menu.createIndexTable(collection.getIndexes());

      expect(result).toMatchSnapshot();
    },
  );
});
