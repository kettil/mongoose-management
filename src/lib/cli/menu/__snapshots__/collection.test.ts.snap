// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the CollectionMenu class it should be return indexes table when createIndexTable() is called with type "index" 1`] = `
Array [
  "[2m|[22m [4mName[24m     [2m|[22m [4mColumn(s)[24m  [2m|[22m [4mUnique[24m [2m|[22m [4mSparse[24m [2m|[22m",
  "[2m|[22m column1_ [2m|[22m column1: 1 [2m|[22m        [2m|[22m        [2m|[22m",
  "[2m|[22m index1   [2m|[22m column1: 1 [2m|[22m        [2m|[22m        [2m|[22m",
  "",
]
`;

exports[`Check the CollectionMenu class it should be return indexes table when createIndexTable() is called with type "sparse" 1`] = `
Array [
  "[2m|[22m [4mName[24m     [2m|[22m [4mColumn(s)[24m  [2m|[22m [4mUnique[24m [2m|[22m [4mSparse[24m [2m|[22m",
  "[2m|[22m column1_ [2m|[22m column1: 1 [2m|[22m        [2m|[22m   âœ”    [2m|[22m",
  "[2m|[22m index1   [2m|[22m column1: 1 [2m|[22m        [2m|[22m        [2m|[22m",
  "",
]
`;

exports[`Check the CollectionMenu class it should be return indexes table when createIndexTable() is called with type "unique" 1`] = `
Array [
  "[2m|[22m [4mName[24m     [2m|[22m [4mColumn(s)[24m  [2m|[22m [4mUnique[24m [2m|[22m [4mSparse[24m [2m|[22m",
  "[2m|[22m column1_ [2m|[22m column1: 1 [2m|[22m   âœ”    [2m|[22m        [2m|[22m",
  "[2m|[22m index1   [2m|[22m column1: 1 [2m|[22m        [2m|[22m        [2m|[22m",
  "",
]
`;

exports[`Check the CollectionMenu class it should be return menu selection when exec() is called 1`] = `
Array [
  Separator {
    "line": "[2mCollection: [0m[1mc1[2m[0m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[4mColumns list[24m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[2m|[2m [4mName[24m      [2m|[2m [4mType[24m     [2m|[2m [4mRequired[24m [2m|[2m [4mDefault[24m   [2m|[2m [4mReference[24m [2m|[2m [4mIndex[24m     [2m|[2m [4mUnique[24m [2m|[2m [4mSparse[24m [2m|[2m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[2m|[2m _id       [2m|[2m objectId [2m|[2m          [2m|[2m           [2m|[2m           [2m|[2m           [2m|[2m        [2m|[2m        [2m|[2m[22m",
    "type": "separator",
  },
  Object {
    "name": "[2m|[22m column1   [2m|[22m string   [2m|[22m          [2m|[22m           [2m|[22m           [2m|[22m           [2m|[22m        [2m|[22m        [2m|[22m",
    "short": "Column - column1",
    "value": Object {
      "data": ColumnDataset {
        "collection": CollectionDataset {
          "columns": Array [
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "_id",
                "type": "objectId",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            [Circular],
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "createdAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "updatedAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
          ],
          "idType": "objectId",
          "indexes": Array [
            IndexDataset {
              "columns": Array [
                Array [
                  [Circular],
                  1,
                ],
              ],
              "index": Object {
                "columns": Object {
                  "column1": 1,
                },
                "name": "index1",
                "properties": Object {},
              },
              "name": "index1",
              "parent": [Circular],
              "properties": Object {},
              "readonly": false,
            },
          ],
          "name": "c1",
          "parent": [MockFunction],
          "specialColumns": Array [
            Array [
              "createdAt",
              "date",
            ],
            Array [
              "updatedAt",
              "date",
            ],
            Array [
              "_id",
              "objectId",
            ],
          ],
        },
        "column": Object {
          "name": "column1",
          "type": "string",
        },
        "columns": Array [],
        "index": undefined,
        "parent": CollectionDataset {
          "columns": Array [
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "_id",
                "type": "objectId",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            [Circular],
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "createdAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "updatedAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
          ],
          "idType": "objectId",
          "indexes": Array [
            IndexDataset {
              "columns": Array [
                Array [
                  [Circular],
                  1,
                ],
              ],
              "index": Object {
                "columns": Object {
                  "column1": 1,
                },
                "name": "index1",
                "properties": Object {},
              },
              "name": "index1",
              "parent": [Circular],
              "properties": Object {},
              "readonly": false,
            },
          ],
          "name": "c1",
          "parent": [MockFunction],
          "specialColumns": Array [
            Array [
              "createdAt",
              "date",
            ],
            Array [
              "updatedAt",
              "date",
            ],
            Array [
              "_id",
              "objectId",
            ],
          ],
        },
        "readonly": false,
        "subTypes": Array [],
      },
    },
  },
  Separator {
    "line": "[2m[2m|[2m createdAt [2m|[2m date     [2m|[2m          [2m|[2m           [2m|[2m           [2m|[2m           [2m|[2m        [2m|[2m        [2m|[2m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[2m|[2m updatedAt [2m|[2m date     [2m|[2m          [2m|[2m           [2m|[2m           [2m|[2m           [2m|[2m        [2m|[2m        [2m|[2m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2mNote: Columns \\"_id\\", \\"createdAt\\" and \\"updatedAt\\" are created automatically[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[4mIndexes list[24m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[2m|[2m [4mName[24m   [2m|[2m [4mColumn(s)[24m  [2m|[2m [4mUnique[24m [2m|[2m [4mSparse[24m [2m|[2m[22m",
    "type": "separator",
  },
  Object {
    "name": "[2m|[22m index1 [2m|[22m column1: 1 [2m|[22m        [2m|[22m        [2m|[22m",
    "short": "Index - index1",
    "value": Object {
      "data": IndexDataset {
        "columns": Array [
          Array [
            ColumnDataset {
              "collection": CollectionDataset {
                "columns": Array [
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "_id",
                      "type": "objectId",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  [Circular],
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "createdAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "updatedAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                ],
                "idType": "objectId",
                "indexes": Array [
                  [Circular],
                ],
                "name": "c1",
                "parent": [MockFunction],
                "specialColumns": Array [
                  Array [
                    "createdAt",
                    "date",
                  ],
                  Array [
                    "updatedAt",
                    "date",
                  ],
                  Array [
                    "_id",
                    "objectId",
                  ],
                ],
              },
              "column": Object {
                "name": "column1",
                "type": "string",
              },
              "columns": Array [],
              "index": undefined,
              "parent": CollectionDataset {
                "columns": Array [
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "_id",
                      "type": "objectId",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  [Circular],
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "createdAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "updatedAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                ],
                "idType": "objectId",
                "indexes": Array [
                  [Circular],
                ],
                "name": "c1",
                "parent": [MockFunction],
                "specialColumns": Array [
                  Array [
                    "createdAt",
                    "date",
                  ],
                  Array [
                    "updatedAt",
                    "date",
                  ],
                  Array [
                    "_id",
                    "objectId",
                  ],
                ],
              },
              "readonly": false,
              "subTypes": Array [],
            },
            1,
          ],
        ],
        "index": Object {
          "columns": Object {
            "column1": 1,
          },
          "name": "index1",
          "properties": Object {},
        },
        "name": "index1",
        "parent": CollectionDataset {
          "columns": Array [
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "_id",
                "type": "objectId",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "column1",
                "type": "string",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": false,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "createdAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "updatedAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
          ],
          "idType": "objectId",
          "indexes": Array [
            [Circular],
          ],
          "name": "c1",
          "parent": [MockFunction],
          "specialColumns": Array [
            Array [
              "createdAt",
              "date",
            ],
            Array [
              "updatedAt",
              "date",
            ],
            Array [
              "_id",
              "objectId",
            ],
          ],
        },
        "properties": Object {},
        "readonly": false,
      },
    },
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
  Separator {
    "line": "[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[22m",
    "type": "separator",
  },
  Object {
    "name": "Create new column",
    "short": "[31mCommand - Create column[39m",
    "value": Object {
      "action": "createColumn",
    },
  },
  Object {
    "name": "Create new index",
    "short": "[31mCommand - Create index[39m",
    "value": Object {
      "action": "createIndex",
    },
  },
  Object {
    "name": "Edit collection",
    "short": "[31mCommand - Edit collection[39m",
    "value": Object {
      "action": "edit",
      "data": undefined,
    },
  },
  Object {
    "name": "Remove collection",
    "short": "[31mCommand - Remove collection[39m",
    "value": Object {
      "action": "remove",
      "data": undefined,
    },
  },
  Object {
    "name": "Back",
    "short": "[31mCommand - Back[39m",
    "value": Object {
      "action": "back",
    },
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
]
`;

exports[`Check the CollectionMenu class it should be return menu selection when getChoiceIndexList() is called with indexes 1`] = `
Array [
  Separator {
    "line": "[2m[2m|[2m [4mName[24m     [2m|[2m [4mColumn(s)[24m  [2m|[2m [4mUnique[24m [2m|[2m [4mSparse[24m [2m|[2m[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m[2m|[2m column1_ [2m|[2m column1: 1 [2m|[2m        [2m|[2m        [2m|[2m[22m",
    "type": "separator",
  },
  Object {
    "name": "[2m|[22m index1   [2m|[22m column1: 1 [2m|[22m        [2m|[22m        [2m|[22m",
    "short": "Index - index1",
    "value": Object {
      "data": IndexDataset {
        "columns": Array [
          Array [
            ColumnDataset {
              "collection": CollectionDataset {
                "columns": Array [
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "_id",
                      "type": "objectId",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  [Circular],
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "createdAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "updatedAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                ],
                "idType": "objectId",
                "indexes": Array [
                  IndexDataset {
                    "columns": Array [
                      Array [
                        [Circular],
                        1,
                      ],
                    ],
                    "index": Object {
                      "columns": Object {
                        "column1": 1,
                      },
                      "name": "column1_",
                      "properties": Object {
                        "sparse": false,
                        "unique": false,
                      },
                      "readonly": true,
                    },
                    "name": "column1_",
                    "parent": [Circular],
                    "properties": Object {
                      "sparse": false,
                      "unique": false,
                    },
                    "readonly": true,
                  },
                  [Circular],
                ],
                "name": "c1",
                "parent": [MockFunction],
                "specialColumns": Array [
                  Array [
                    "createdAt",
                    "date",
                  ],
                  Array [
                    "updatedAt",
                    "date",
                  ],
                  Array [
                    "_id",
                    "objectId",
                  ],
                ],
              },
              "column": Object {
                "name": "column1",
                "type": "string",
              },
              "columns": Array [],
              "index": IndexDataset {
                "columns": Array [
                  Array [
                    [Circular],
                    1,
                  ],
                ],
                "index": Object {
                  "columns": Object {
                    "column1": 1,
                  },
                  "name": "column1_",
                  "properties": Object {
                    "sparse": false,
                    "unique": false,
                  },
                  "readonly": true,
                },
                "name": "column1_",
                "parent": CollectionDataset {
                  "columns": Array [
                    ColumnDataset {
                      "collection": [Circular],
                      "column": Object {
                        "name": "_id",
                        "type": "objectId",
                      },
                      "columns": Array [],
                      "index": undefined,
                      "parent": [Circular],
                      "readonly": true,
                      "subTypes": Array [],
                    },
                    [Circular],
                    ColumnDataset {
                      "collection": [Circular],
                      "column": Object {
                        "name": "createdAt",
                        "type": "date",
                      },
                      "columns": Array [],
                      "index": undefined,
                      "parent": [Circular],
                      "readonly": true,
                      "subTypes": Array [],
                    },
                    ColumnDataset {
                      "collection": [Circular],
                      "column": Object {
                        "name": "updatedAt",
                        "type": "date",
                      },
                      "columns": Array [],
                      "index": undefined,
                      "parent": [Circular],
                      "readonly": true,
                      "subTypes": Array [],
                    },
                  ],
                  "idType": "objectId",
                  "indexes": Array [
                    [Circular],
                    [Circular],
                  ],
                  "name": "c1",
                  "parent": [MockFunction],
                  "specialColumns": Array [
                    Array [
                      "createdAt",
                      "date",
                    ],
                    Array [
                      "updatedAt",
                      "date",
                    ],
                    Array [
                      "_id",
                      "objectId",
                    ],
                  ],
                },
                "properties": Object {
                  "sparse": false,
                  "unique": false,
                },
                "readonly": true,
              },
              "parent": CollectionDataset {
                "columns": Array [
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "_id",
                      "type": "objectId",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  [Circular],
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "createdAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "updatedAt",
                      "type": "date",
                    },
                    "columns": Array [],
                    "index": undefined,
                    "parent": [Circular],
                    "readonly": true,
                    "subTypes": Array [],
                  },
                ],
                "idType": "objectId",
                "indexes": Array [
                  IndexDataset {
                    "columns": Array [
                      Array [
                        [Circular],
                        1,
                      ],
                    ],
                    "index": Object {
                      "columns": Object {
                        "column1": 1,
                      },
                      "name": "column1_",
                      "properties": Object {
                        "sparse": false,
                        "unique": false,
                      },
                      "readonly": true,
                    },
                    "name": "column1_",
                    "parent": [Circular],
                    "properties": Object {
                      "sparse": false,
                      "unique": false,
                    },
                    "readonly": true,
                  },
                  [Circular],
                ],
                "name": "c1",
                "parent": [MockFunction],
                "specialColumns": Array [
                  Array [
                    "createdAt",
                    "date",
                  ],
                  Array [
                    "updatedAt",
                    "date",
                  ],
                  Array [
                    "_id",
                    "objectId",
                  ],
                ],
              },
              "readonly": false,
              "subTypes": Array [],
            },
            1,
          ],
        ],
        "index": Object {
          "columns": Object {
            "column1": 1,
          },
          "name": "index1",
          "properties": Object {},
        },
        "name": "index1",
        "parent": CollectionDataset {
          "columns": Array [
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "_id",
                "type": "objectId",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "column1",
                "type": "string",
              },
              "columns": Array [],
              "index": IndexDataset {
                "columns": Array [
                  Array [
                    [Circular],
                    1,
                  ],
                ],
                "index": Object {
                  "columns": Object {
                    "column1": 1,
                  },
                  "name": "column1_",
                  "properties": Object {
                    "sparse": false,
                    "unique": false,
                  },
                  "readonly": true,
                },
                "name": "column1_",
                "parent": [Circular],
                "properties": Object {
                  "sparse": false,
                  "unique": false,
                },
                "readonly": true,
              },
              "parent": [Circular],
              "readonly": false,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "createdAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
            ColumnDataset {
              "collection": [Circular],
              "column": Object {
                "name": "updatedAt",
                "type": "date",
              },
              "columns": Array [],
              "index": undefined,
              "parent": [Circular],
              "readonly": true,
              "subTypes": Array [],
            },
          ],
          "idType": "objectId",
          "indexes": Array [
            IndexDataset {
              "columns": Array [
                Array [
                  ColumnDataset {
                    "collection": [Circular],
                    "column": Object {
                      "name": "column1",
                      "type": "string",
                    },
                    "columns": Array [],
                    "index": [Circular],
                    "parent": [Circular],
                    "readonly": false,
                    "subTypes": Array [],
                  },
                  1,
                ],
              ],
              "index": Object {
                "columns": Object {
                  "column1": 1,
                },
                "name": "column1_",
                "properties": Object {
                  "sparse": false,
                  "unique": false,
                },
                "readonly": true,
              },
              "name": "column1_",
              "parent": [Circular],
              "properties": Object {
                "sparse": false,
                "unique": false,
              },
              "readonly": true,
            },
            [Circular],
          ],
          "name": "c1",
          "parent": [MockFunction],
          "specialColumns": Array [
            Array [
              "createdAt",
              "date",
            ],
            Array [
              "updatedAt",
              "date",
            ],
            Array [
              "_id",
              "objectId",
            ],
          ],
        },
        "properties": Object {},
        "readonly": false,
      },
    },
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
]
`;

exports[`Check the CollectionMenu class it should be return menu selection when getChoiceIndexList() is called without indexes 1`] = `
Array [
  Separator {
    "line": "[2m- No indexes defined -[22m",
    "type": "separator",
  },
  Separator {
    "line": "[2m [22m",
    "type": "separator",
  },
]
`;
